qiskit_nature/operators/second_quantization/spin_op.py

```diff
@@ -12,117 +12,469 @@
+    _DENSE_LABEL_PATTERN = re.compile(r"^[IXYZ]+$")
+
+    def __init__(
+        self,
+        data: Union[
+            str,
+            List[Tuple[str, complex]],
+            Tuple[np.ndarray, np.ndarray],
+        ],
+        spin: float = 1 / 2,
+    ):
+        r"""Initialize ``SpinOp``.
+
+        Args:
+            data: label string or list of labels and coefficients. See documentation of SpinOp for
+                  more details.
+            spin: positive integer or half-integer which represents spin.
+            dtype: data type of coefficients.
+
+        Raises:
+            ValueError: invalid data is given.
+            QiskitNatureError: invalid spin value
+        """
+        self._coeffs: np.ndarray
+        self._spin_array: np.ndarray
+        dtype = np.complex128  # TODO: configurable data type. mixin?
+
+        if (round(2 * spin) != 2 * spin) or (spin <= 0):
+            raise QiskitNatureError("spin must be a positive integer or half-integer")
+        self._dim = int(round(2 * spin)) + 1
+
+        if isinstance(data, tuple):
+            self._spin_array = np.array(data[0], dtype=np.uint8)
+            self._register_length = self._spin_array.shape[2]
+            self._coeffs = np.array(data[1], dtype=dtype)
 
-        # 2. Initialize the operator list of `self`
-        self._operator_list = operator_list
+        if isinstance(data, str):
+            data = [(data, 1)]
 
-        # 3. Set the operators particle type to 'spin S' with S the spin value (as float with 1
-        # decimal).
-        # SumOp.__init__(self, particle_type='spin {0:.1f}'.format(self.spin))
+        if isinstance(data, list):
+            if not all(self._VALID_LABEL_PATTERN.match(label) for label, _ in data):
+                raise ValueError(f"Invalid label are included in {data}")
+
+            data = self._flatten_ladder_ops(data)
+
+            labels, coeffs = zip(*data)
+            self._coeffs = np.array(coeffs, dtype=dtype)
+
+            if all(self._SPARSE_LABEL_PATTERN.match(label) for label in labels):
+                self._from_sparse_label(labels)
+            elif all(self._DENSE_LABEL_PATTERN.match(label) for label in labels):
+                self._register_length = len(labels[0])
+                self._spin_array = np.array(
+                    [
+                        [[char == "X", char == "Y", char == "Z"] for char in label]
+                        for label in labels
+                    ],
+                    dtype=np.uint8,
+                ).transpose((2, 0, 1))
+            else:
+                raise ValueError(f"Invalid labels or mixed labels are included in {labels}")
 
     @property
     def register_length(self):
         return self._register_length
 
-    def __repr__(self):
-        full_str = ''
-        for operator in self._operator_list:
-            full_str += '{1} \t {0}\n'.format(operator.coeff, operator.label)
-        return full_str
+    @property
+    def spin(self) -> float:
```

Comment:
How about returning `Fraction ` here? (but I think `float` is well.)

Reply:
Since this is always an integral number of half-spins I guess that is why I suggesting the multiplier. So in my example above, if you gave it 7 in the constructor (for 7/2) you would get 7 back here. This could then be an integer without any rounding/precision issues. Maybe since this is dividing by 2 there would not be any precision error. But seeing round() and things in the code leads me to believe such precision issues are being catered to. Maybe its not a big deal and float is fine.

Reply:
I don't think going to an integer value is a user-friendly decision here. Instead, we discussed offline that the `Fraction`-approach is probably best suited. Please see also my proposed changes below.

Reply:
Chemists are used to multiplicity etc. but I understand having a multiplier is not quite as nice as specifying the actual spin value. Fraction will let you specify things so you avoid any precision/rounding etc that might occur/be needed.